support/mesh/mesh_channel_core/src/deque.rs

```diff
@@ -0,0 +1,359 @@
+            offset - self.cap
+        } else {
+            offset
+        };
+        debug_assert!(offset + self.vtable.layout.size() <= self.cap);
+        offset
+    }
+
+    /// # Safety
+    /// The caller must ensure that `i` is in bounds.
+    unsafe fn buf_at(&mut self, i: usize) -> InPlaceElement<'_> {
+        let offset = self.offset(i);
+        // SAFETY: `i` is a valid index into `buf`.
+        let ptr = unsafe { self.buf.add(offset).cast() };
+        InPlaceElement(ptr, PhantomData)
+    }
+
+    /// Reserves space for one element and returns a pointer to it.
+    pub fn reserve_one(&mut self) -> InPlaceElement<'_> {
+        if self.len >= self.cap {
+            self.grow();
+        }
+        // SAFETY: We just grew the buffer if necessary.
+        unsafe { self.buf_at(self.len) }
+    }
+
+    /// Commits the next element. Used after calling `reserve_one` and writing
+    /// to the buffer.
+    ///
+    /// # Safety
+    /// The caller must ensure that the element has been written to the buffer.
+    pub unsafe fn commit_one(&mut self) {
+        self.len += self.vtable.element_len;
+        debug_assert!(self.len <= self.cap);
+    }
+
+    /// Pushes a new element to the back of the queue.
+    ///
+    /// # Safety
+    /// The caller must ensure that `value` is a valid owned pointer to the
+    /// element type that this queue was created with.
+    ///
+    /// Additionally, once at least one element has been pushed to this
+    /// queue, the caller must ensure that the queue is not sent/shared across
+    /// threads unless the element type is `Send` and `Sync`.
+    pub unsafe fn push_back(&mut self, value: *const ()) {
+        let dst = self.reserve_one();
+        // SAFETY: the caller ensures that `value` is a valid owned pointer to the
+        // element type.
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                value.cast(),
+                dst.as_ptr().cast::<u8>(),
+                self.vtable.layout.size(),
+            );
+            self.commit_one();
+        }
+    }
+
+    /// Pops the front element from the queue and returns a pointer to it.
+    ///
+    /// The caller is responsible for taking ownership of the element to ensure
+    /// that it is properly dropped.
+    pub fn pop_front_in_place(&mut self) -> Option<InPlaceElement<'_>> {
+        if self.len == 0 {
+            return None;
+        };
+        let head = self.head;
+        self.head = self.offset(self.vtable.layout.size());
+        self.len -= self.vtable.element_len;
+        // SAFETY: `head` is a valid index into `buf`.
+        let ptr = unsafe { self.buf.add(head).cast() };
+        Some(InPlaceElement(ptr, PhantomData))
+    }
+
+    /// Pops the front element from the queue and writes it to `dst`.
+    ///
+    /// # Safety
+    /// The caller must ensure that `dst` is valid for writing the element.
+    pub unsafe fn pop_front(&mut self, dst: *mut ()) {
```

Comment:
can this API be improved by having `dst` be a MaybeUninit, and then returning an initialized ptr?

Reply:
`MaybeUninit<T>` of what `T`?

Reply:
ah, well, yeah. shoot. ok, nvmd

Reply:
well, that is, unless you want to make this method generic over T

Reply:
The caller doesn't know `T` either.

Reply:
that's what I get for leaving comments without reading the second half of the PR yet 🙃